Add DeferRunInterval to Improve Intune Retry Handling and User Experience with Show-ADTInstallationWelcome Prompt

tjgruber opened the pull request:
#### **Overview**
This PR introduces a new parameter, `-DeferRunInterval`, to the `PSAppDeployToolkit` (`Show-ADTInstallationWelcome` and related functions). The feature aims to mitigate repetitive installation attempts by Intune after a user defers an installation, enhancing the overall user experience.

This is a minimally invasive but effective way that addresses the issue.

#### **Key Changes**
1. **New Parameter Added:**
   - `DeferRunInterval` (as **_System.TimeSpan_**): Specifies the interval before re-prompting the user after a deferral.
   - Improves user experience by avoiding multiple immediate retries caused by Intune's default behavior.
   - Seamlessly integrates with the existing deferral workflow in `Show-ADTInstallationWelcome`.

2. **Updated Functions:**
   - **`Set-ADTDeferHistory`**: Supports saving and retrieving the `DeferRunInterval` and the last deferral timestamp (`DeferRunIntervalLastTime`).
   - **`Show-ADTInstallationWelcome`**: Logic added to:
     - Check if the deferral interval has elapsed.
     - Exit gracefully if the interval has not yet elapsed.

3. **Changes in `PSADT.Module.Class.DeploymentSession.cs`:**
   - Added logic to handle the new parameters: `DeferRunInterval` and `DeferRunIntervalLastTime`.

#### **Test Cases**
All test scenarios were validated successfully:
1. **With `DeferRunInterval` Specified**:
   - Prevents repeated prompts during the specified interval.
2. **Without `DeferRunInterval` Specified**:
   - Falls back to default deferral logic without affecting the user experience.
3. **Emergency Update Scenario**:
   - If `-DeferRunInterval` is not specified in a subsequent deployment, existing entries in the registry do not interfere with prompting users.
4. **TestLocal**:
   - The 'TestLocal' test in VSCode passed successfully.
5. **Local Build**:
   - Module built and tested locally with expected results successfully.

#### **Behavioral Improvements**
- Enhanced flexibility for deployments managed through Intune.
- Significant reduction in redundant prompts, leading to better user satisfaction.

#### **Log Outputs**
Clear and actionable log messages added to track `DeferRunInterval` behavior:
- Recording of `DeferRunInterval` and `DeferRunIntervalLastTime`.
- Logging checks for elapsed intervals and graceful exits.

#### **Backward Compatibility**
The changes are fully backward-compatible:
- If `-DeferRunInterval` is not specified, the toolkit functions as before.

---

Let me know if you'd like further tweaks or have any questions!

---

Examples used for logs below:
```powershell
Show-ADTInstallationWelcome -CloseProcesses 'code' -Subtitle 'Application Update' -AllowDeferCloseProcesses -CheckDiskSpace -ForceCloseProcessesCountdown 3270 -NoMinimizeWindows -DeferDays 5 -DeferRunInterval ([System.TimeSpan]::FromHours(24))

Show-ADTInstallationWelcome -CloseProcesses 'code' -Subtitle 'Application Update' -AllowDeferCloseProcesses -CheckDiskSpace -ForceCloseProcessesCountdown 3270 -NoMinimizeWindows -DeferDays 5
```

---

Logging examples:
```
# Test Case: (First run with `-DeferRunInterval ([System.TimeSpan]::FromHours(24))` specified)
<![LOG[[Pre-Install] :: The following processes are running: [Code].]LOG]!>
<![LOG[[Pre-Install] :: DeferRunInterval of [1.00:00:00] is specified. Checking DeferRunIntervalLastTime.]LOG]!>
<![LOG[[Pre-Install] :: Installation deferred by the user.]LOG]!>
<![LOG[[Pre-Install] :: Converting the date [2025-01-13 12:30:52] to a universal sortable date time pattern based on the current culture [en-US].]LOG]!>
<![LOG[[Pre-Install] :: Setting deferral history: [DeferTimesRemaining = 0].]LOG]!>
<![LOG[[Pre-Install] :: Setting deferral history: [DeferDeadline = 2025-01-18 12:30:46Z].]LOG]!>
<![LOG[[Pre-Install] :: Setting deferral history: [DeferRunInterval = 1.00:00:00].]LOG]!>
<![LOG[[Pre-Install] :: Setting deferral history: [DeferRunIntervalLastTime = 2025-01-13 20:30:52Z].]LOG]!>
<![LOG[[Finalization] :: [PSAppDeployToolkit_4.1.0_EN_01] install completed with exit code [60012].]LOG]!>


# Test Case: (Immediate subsequent run mimicking Intune retry after user deferral)
<![LOG[[Pre-Install] :: The following processes are running: [Code].]LOG]!>
<![LOG[[Pre-Install] :: DeferRunInterval of [1.00:00:00] is specified. Checking DeferRunIntervalLastTime.]LOG]!>
<![LOG[[Pre-Install] :: Converting the date [2025-01-13 20:30:52Z] to a universal sortable date time pattern based on the current culture [en-US].]LOG]!>
<![LOG[[Pre-Install] :: Converting the date [2025-01-14 12:30:52Z] to a universal sortable date time pattern based on the current culture [en-US].]LOG]!>
<![LOG[[Pre-Install] :: Converting the date [2025-01-13 12:31:08] to a universal sortable date time pattern based on the current culture [en-US].]LOG]!>
<![LOG[[Pre-Install] :: DeferRunInterval has not elapsed. Exiting gracefully.]LOG]!>
<![LOG[[Finalization] :: [PSAppDeployToolkit_4.1.0_EN_01] install completed with exit code [1618].]LOG]!>


# Test Case: (User deferral without `-DeferRunInterval ([System.TimeSpan]::FromHours(24))` specified)
<![LOG[[Pre-Install] :: The following processes are running: [Code].]LOG]!>
<![LOG[[Pre-Install] :: Installation deferred by the user.]LOG]!>
<![LOG[[Pre-Install] :: Setting deferral history: [DeferTimesRemaining = 0].]LOG]!>
<![LOG[[Pre-Install] :: Setting deferral history: [DeferDeadline = 2025-01-18 12:30:46Z].]LOG]!>
<![LOG[[Finalization] :: [PSAppDeployToolkit_4.1.0_EN_01] install completed with exit code [60012].]LOG]!>


# Test Case: (`-DeferRunInterval ([System.TimeSpan]::FromHours(24))` is specified but process is not running)
<![LOG[[Pre-Install] :: Specified applications are not running.]LOG]!>
<![LOG[[Pre-Install] :: Creating the progress dialog in a separate thread with message: [Installation in progress. Please wait...].]LOG]!>
<![LOG[[Post-Install] :: Displaying custom installation prompt asynchronously with the parameters: [-Message:'You can customize...].]LOG]!>
<![LOG[[Finalization] :: Closing the installation progress dialog.]LOG]!>
<![LOG[[Finalization] :: Removing deferral history...]LOG]!>
<![LOG[[Finalization] :: [PSAppDeployToolkit_4.1.0_EN_01] install completed with exit code [0].]LOG]!>
```

mjr4077au commented on src/PSADT/PSADT/Module/PSADT.Module.Class.DeploymentSession.cs:
```diff
@@ -1076,7 +1076,9 @@ private void CreateDeferHistoryPath()
-        public void SetDeferHistory(int? deferTimesRemaining, string deferDeadline)
+        /// <param name="deferRunInterval">The interval in minutes before prompting the user again after a deferral.</param>
+        /// <param name="deferRunIntervalLastTime">The timestamp of the last deferRunInterval.</param>
+        public void SetDeferHistory(int? deferTimesRemaining, string deferDeadline, int? deferRunInterval = null, string deferRunIntervalLastTime = null)
```
The new string should be nullable based on the setup performed on the PowerShell side.

mjr4077au commented on src/PSAppDeployToolkit/Public/Set-ADTDeferHistory.ps1:
```diff
@@ -54,12 +61,16 @@ function Set-ADTDeferHistory
     try
     {
-        (Get-ADTSession).SetDeferHistory($(if ($PSBoundParameters.ContainsKey('DeferTimesRemaining')) { $DeferTimesRemaining }), $DeferDeadline)
+        (Get-ADTSession).SetDeferHistory($(if ($PSBoundParameters.ContainsKey('DeferTimesRemaining')) { $DeferTimesRemaining }), $DeferDeadline, $(if ($PSBoundParameters.ContainsKey('DeferRunInterval')) { $DeferRunInterval }), $(if ($PSBoundParameters.ContainsKey('DeferRunInterval')) { (Get-ADTUniversalDate) }))
```
While it's not overly expensive, we're doing a dictionary check twice with `$PSBoundParameters.ContainsKey('DeferRunInterval')`. Can we either cache the result, or do something like:
```PowerShell
if ($PSBoundParameters.ContainsKey('DeferRunInterval'))
{
    (Get-ADTSession).SetDeferHistory($(if ($PSBoundParameters.ContainsKey('DeferTimesRemaining')) { $DeferTimesRemaining }), $DeferDeadline, $DeferRunInterval, (Get-ADTUniversalDate))
}
else
{
    (Get-ADTSession).SetDeferHistory($(if ($PSBoundParameters.ContainsKey('DeferTimesRemaining')) { $DeferTimesRemaining }), $DeferDeadline, $null, $null)
}

mjr4077au commented on src/PSAppDeployToolkit/Public/Show-ADTInstallationWelcome.ps1:
```diff
@@ -191,6 +196,10 @@ function Show-ADTInstallationWelcome
         [ValidateNotNullOrEmpty()]
         [System.String]$DeferDeadline,
 
+        [Parameter(Mandatory = $false, HelpMessage = 'Specifies the number of minutes that must elapse before prompting the user again if a process listed in "CloseProcesses" is still running after a deferral.')]
```
I'm not keen on some parameters using seconds and others using minutes. I think the units should be the same for both. It's regrettable that we didn't change CloseProcessesCountdown to a TimeSpan object as that would be preferable anywhere we need to represent the passage of time.

Can you please make this a System.TimeSpan object? Once all review items of the PR have been performed and reviewed by myself and a colleague, I'll look to making the CloseProcessesCountdown et al be System.TimeSpan parameters also and will phase out the usage of seconds.

tjgruber replied:
Changed to System.TimeSpan and updated param info appropriately.

mjr4077au reviewed (changes requested):
All in all, it looks pretty good. Once the few minor things are addressed, I'll ask @DanGough  to have a peer review also.

tjgruber commented:
> All in all, it looks pretty good. Once the few minor things are addressed, I'll ask @DanGough to have a peer review also.

@mjr4077au Thanks for the feedback! I believe I addressed all the points you mentioned, including the adjustments to the affected areas. Let me know if there's anything else that needs tweaking or if you'd like me to clarify or refine further before it goes for peer review.

mjr4077au commented on src/PSADT/PSADT/Module/PSADT.Module.Class.DeploymentSession.cs:
```diff
@@ -1099,6 +1101,24 @@ public void SetDeferHistory(int? deferTimesRemaining, string deferDeadline)
                 }
                 moduleSessionState.InvokeProvider.Property.New([RegKeyDeferHistory], "DeferDeadline", "String", deferDeadline, true, true);
             }
+            if (null != deferRunInterval)
```
@tjgruber I might be wrong, but do we need to archive off DeferRunInterval at all? I can't see it being retrieved and used anywhere (I am on my phone though).

tjgruber replied:
It's not being retrieved from the registry, but is stored there for reference with the last deferred time so it can be known just by looking at the deferral history what it was set to at that time. It also leaves the option open to be used elsewhere if/when needed.

tjgruber replied:
I suppose to answer your question directly; it's not a need and can safely be removed from archival without impacting anything currently. I personally liked it there when looking at the defer history in the registry so I could be sure things were happening correctly.

mjr4077au replied:
No worries, that's good enough for me

tjgruber commented:
@mjr4077au I was testing smaller time spans such as 1 minute, and discovered I missed converting the `DeferRunIntervalNextTime` to universal time. Fixed that and now all time spans result in the expected behavior.

mjr4077au commented:
This is taking too long to get reviewed and I've got some pending work in `Show-ADTInstallationWelcome`, so I'm merging it in for now so I can continue onwards with things.

Thanks for your contribution! 🤘

DanGough commented:
Looks good. I'd like to add a few more examples to help though to show alternative ways of defining the timespan, as the PowerShell native way may be more user friendly, e.g:

New-TimeSpan -Hours 4
New-TimeSpan -Days 1
